### PR TITLE
feat(rust): implement channel tunneling, update ockamd to use it

### PR DESCRIPTION
**Pull request by robinbudd**

### CODE REVIEW

File: implementations/rust/daemon/src/cli.rs [hunk 1]
```diff
@@ -37,15 +37,21 @@ pub struct Args {
         default_value = "stdout",
         help = r#"Route to channel responder, e.g. udp://host:port[,udp://host:port] (note comma-separation) or "stdout""#
     )]
-    route: OutputKind,
+    sink_route: OutputKind,
```
Comment (nilslice) [inline implementations/rust/daemon/src/cli.rs:40]: this breaks the CLI -- i don't think we need to disambiguate the `sink_route` to end-users, as (I believe) there's only 1 kind of `route`.

File: implementations/rust/daemon/src/cli.rs [hunk 2]
```diff
@@ -241,19 +257,23 @@ impl FromStr for VaultKind {
 pub enum ChannelRole {
     /// The Initiator role expects a channel responder address and a public key to use in order to
     /// communicate with the Responder end of the channel.
-    Initiator,
+    Source,
     /// The Responder role will create a channel responder, and will instruct the program to print
```
Comment (nilslice) [inline implementations/rust/daemon/src/cli.rs:261]: If we're going to commit to using these names in code (which I think @mrinalwadhwa should approve) then can you also update the comments to reflect the change?

File: implementations/rust/daemon/src/cli.rs [hunk 3]
```diff
@@ -241,19 +257,23 @@ impl FromStr for VaultKind {
...
 impl FromStr for ChannelRole {
     type Err = String;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         match s {
-            "initiator" | "init" => Ok(ChannelRole::Initiator),
-            "responder" | "resp" => Ok(ChannelRole::Responder),
+            "initiator" | "init" => Ok(ChannelRole::Source),
+            "responder" | "resp" => Ok(ChannelRole::Sink),
+            "router" => Ok(ChannelRole::Router),
```
Comment (nilslice) [inline implementations/rust/daemon/src/cli.rs:276]: these should also reflect the naming change (again, if we're going to commit to it & @mrinalwadhwa approves).. this particular function takes the string literal from the CLI input for the `--role` arg, and translates it into the proper enum variant. So we want these to be consistent as well.

File: implementations/rust/daemon/src/config.rs [hunk 4]
```diff
@@ -110,8 +123,9 @@ impl From<cli::Args> for Config {
         }
 
         cfg.role = match args.role() {
-            cli::ChannelRole::Initiator => Role::Initiator,
-            cli::ChannelRole::Responder => Role::Responder,
+            cli::ChannelRole::Source => Role::Initiator,
+            cli::ChannelRole::Sink => Role::Responder,
+            cli::ChannelRole::Router => Role::Router,
```
Comment (nilslice) [inline implementations/rust/daemon/src/config.rs:128]: Same consistency suggestion as before -- I don't think we should have different names between internal / external concerns (e.g. it's all external)

File: implementations/rust/daemon/src/initiator.rs [hunk 5]
```diff
@@ -24,9 +26,31 @@ pub fn run(config: Config) {
...
+    }
+    if let Some(channel_to_sink) = config.channel_to_sink() {
+        onward_route
+            .addresses
+            .push(RouterAddress::channel_router_address_from_str(&channel_to_sink).unwrap());
+    }
+    onward_route
+        .addresses
+        .push(RouterAddress::channel_router_address_from_str(CHANNEL_ZERO).unwrap());
+    println!("Initiating channel with route");
```
Comment (nilslice) [inline implementations/rust/daemon/src/initiator.rs:48]: let's remove this println

File: implementations/rust/daemon/src/node.rs [hunk 6]
```diff
@@ -128,12 +128,14 @@ impl<'a> Node<'a> {
             )))
             .expect("failed to register worker with router");
 
+        println!("worker registered");
```
Comment (nilslice) [inline implementations/rust/daemon/src/node.rs:131]: lets remove this println

File: implementations/rust/daemon/src/node.rs [hunk 7]
```diff
@@ -128,12 +128,14 @@ impl<'a> Node<'a> {
...
         self.worker = Some(worker);
     }
 
     pub fn run(mut self) {
         match self.worker {
-            Some(worker) => {
+            Some(mut worker) => {
+                println!("will poll worker");
```
Comment (nilslice) [inline implementations/rust/daemon/src/node.rs:138]: lets remove this println

File: implementations/rust/daemon/src/responder.rs [hunk 8]
```diff
@@ -4,14 +4,35 @@ use crate::config::{AddonKind, Config};
...
     let (mut node, router_tx) = Node::new(&config);
 
     let worker_addr = RouterAddress::worker_router_address_from_str("01242020").unwrap();
+    // kick off secure channel to router, if we have a router address
+    match config.router_socket() {
+        Some(socket) => {
+            let route = Route {
+                addresses: vec![
+                    RouterAddress::from_address(Address::UdpAddress(socket)).unwrap(),
+                    RouterAddress::channel_router_address_from_str("00000000").unwrap(),
```
Comment (nilslice) [inline implementations/rust/daemon/src/responder.rs:22]: can this be replaced with the CHANNEL_ZERO constant?

File: implementations/rust/daemon/src/worker.rs [hunk 9]
```diff
@@ -52,7 +57,24 @@ impl Worker {
...
+        //     if rpk == encode(&m.message_body) {
+        //         println!("keys agree");
+        //         return Ok(());
+        //     } else {
+        //         println!("keys conflict");
+        //         return Err("remote public key doesn't match expected, possible spoofing".into());
+        //     }
+        // }
+        // Ok(())
+    }
```
Comment (nilslice) [inline implementations/rust/daemon/src/worker.rs:75]: does this introduce a regression? are we not validating the public key elsewhere if this is commented out?

File: implementations/rust/daemon/src/cli.rs (hunk 3 above)
Comment (mrinalwadhwa) [reply implementations/rust/daemon/src/cli.rs:276]: Please help me understand the thought behind this?

**Review by nilslice — CHANGES_REQUESTED** [review]
Need some updates before it can be tested - namely the identifier name change for Source/Sink and making that consistent, and the approval from @mrinalwadhwa that we want to do so. 

Also, the CLI arg name change should probably be reverted so that the existing docker demo and the slide in mrinal's Influx deck is still valid.

File: implementations/rust/daemon/src/cli.rs (hunk 3 above)
Comment (robinbudd) [reply implementations/rust/daemon/src/cli.rs:276]: @mrinalwadhwa The "responder" role is no longer really appropriate for the sink end of the pipe (i.e. the end that pumps data into Influx). The sink acts as initiator when it establishes the channel with the router, and as responder when it gets M1 from the source end of the pipe. "Source" and "sink" are less ambiguous IMO and don't muddy the initiator/responder relationship.

File: implementations/rust/daemon/src/cli.rs (hunk 1 above)
Comment (robinbudd) [reply implementations/rust/daemon/src/cli.rs:40]: ok

File: implementations/rust/daemon/src/cli.rs (hunk 2 above)
Comment (robinbudd) [reply implementations/rust/daemon/src/cli.rs]: done...

File: implementations/rust/daemon/src/cli.rs (hunk 3 above)
Comment (robinbudd) [reply implementations/rust/daemon/src/cli.rs]: done

File: implementations/rust/daemon/src/initiator.rs (hunk 5 above)
Comment (robinbudd) [reply implementations/rust/daemon/src/initiator.rs]: done

File: implementations/rust/daemon/src/node.rs (hunk 7 above)
Comment (robinbudd) [reply implementations/rust/daemon/src/node.rs]: done

File: implementations/rust/daemon/src/worker.rs (hunk 9 above)
Comment (robinbudd) [reply implementations/rust/daemon/src/worker.rs:75]: This is the sink, which publishes its public key. It's the source side that validates the public key.

**Comment by nilslice** [comment]
All of this is rolled into #620, so I'm closing this PR in favor of that.

**Pull request closed without merging**